File: Scripts/2024/DH_to_JA.m
function [Ja] = DH_to_JA(DHTABLE, prismatic_indices)
    %Function that find the angular jacobian for a robot 
    %
    %parameters:
    %-DH in the order alpha a d theta, ex:
    %   DHTABLE = [ pi/2    0       sym('d1')   q1;
    %               pi/2    0           0       q2;
    %               pi/2    0          q3       0;
    %               0      sym('a4')    0       q4];
    %-prismatic_indices: a list of indices of prismatic Joints ex: [2,4]
    %   if there is no prismatic joint put []
    syms alpha_ d a_ theta_

    N = size(DHTABLE, 1);

    TDH = [cos(theta_) -sin(theta_)*cos(alpha_) sin(theta_)*sin(alpha_) a_*cos(theta_);
           sin(theta_) cos(theta_)*cos(alpha_) -cos(theta_)*sin(alpha_) a_*sin(theta_);
           0 sin(alpha_) cos(alpha_) d;
           0 0 0 1];

    A = cell(1, N);

    
    for i = 1:N
        A{i} = subs(TDH, [alpha_, a_, d, theta_], DHTABLE(i, :));
    end

    T = eye(4);
    rotation_0_to_i = cell(1, N);

    for i = 1:N
        T = T * A{i};
        rotation_0_to_i{i} = simplify(T);
    end

    % Initialize the Jacobian matrix as symbolic
    Ja = sym(zeros(3, N));

    % Unit vector along the z-axis
    z = [0; 0; 1];
    Ja(:, 1) = z;
    % Compute the column vectors for the Jacobian matrix
    for i = 1:N-1
        R = rotation_0_to_i{i}(1:3, 1:3);
        zi = R * z;
        Ja(:, i+1) = zi;
    end

    % Substitute columns with zeros for prismatic joints
    for prismatic_index = prismatic_indices
        if prismatic_index ~= N
            Ja(:, prismatic_index) = zeros(3, 1);
        end
    end

    % Display or return the Jacobian matrix as needed
    disp('Angular Jacobian Matrix:');
    disp(Ja);
end

File: Scripts/2025/python_scripts/robobox/DH_to_JA.py
import sympy as sp
from .DH_matrix import DH_matrix  

def DH_to_JA(DHTABLE: list, prismatic_joints: list=[]) -> sp.Matrix:
    """
    Computes the Angular Jacobian matrix from the provided DH table
    
    @param DHTABLE: The Denavit-Hartenberg table
    @param prismatic_joints: The list of prismatic joint indexes
    """
    N = len(DHTABLE)
    
    alpha, a, d, theta = sp.symbols('alpha a d theta')
    
    DH_mat = DH_matrix()

    A = []
    for i in range(N):
        A_i = DH_mat.subs({
            alpha: DHTABLE[i][0],
            a: DHTABLE[i][1],
            d: DHTABLE[i][2],
            theta: DHTABLE[i][3]
        })
        A.append(A_i)

    T = sp.eye(4)
    R0i = []

    for i in range(N):
        T = T @ A[i]
        R0i.append(T[0:3, 0:3])

    JA = sp.zeros(3, N)

    z = sp.Matrix([0, 0, 1])
    JA[:, 0] = z

    for i in range(1, N):
        R = R0i[i-1]   # Rotation matrix for joint i-1
        zi = R @ z     # Transformed z-axis
        JA[:, i] = zi

    # Remove the prismatic joints
    if len(prismatic_joints) > 0:
        for prismatic_index in prismatic_joints:
            if prismatic_index != N:
                JA[:, prismatic_index] = sp.zeros(3, 1)

    return JA